### PR TITLE
8.0.x

### PR DESCRIPTION
**Pull request by amaria**
This next round of changes, attaches the Add/Edit buttons to entity_reference form fields. The buttons don't actually do anything yet but you should be able to see them on fields where the settings are enabled.
This commit....
- Adds 'entityconnect_submit' element type for the buttons.
- Uses hook_form_alter() to attach a process callback to entity_reference field widgets.
- Adds a class (EntityconnectWidgetProcessor) to handle the process callback which actually adds the button elements.
- General code cleanup.

### CODE REVIEW

**Comment by amaria** [comment]
Oh, forgot to mention that I didn't test this with a view list yet so not sure if it works. Probably not.

**Comment by amaria** [comment]
Now this works with Views reference types plus I fixed some problems with the button insertions so now appear correctly in all widgets/cardinality/multiselection scenarios.

There is still a bug when clicking the "Add Another item" button on the Autocomplete widget, that I'll look at later.

**Comment by amaria** [comment]
More changes...
1. Fixed the bug when clicking other submit buttons.
2. Moved the form alter code to a new class in order to streamline the module file.

I think it's stable now. :)

**Comment by jygastaud** [comment]
Thanks a lot @amaria 
I will try to give a try ASAP

**Comment by jygastaud** [comment]
Hi and thanks again for your amazing work,

I give your MR a (really) quick try on Drupal 8.0.5 using SQLite.
I created a reference field for content, with Article and Page allowed added on Article content type.

Buttons display looks to work great !

However I have some issues :
1. When I "Add" a reference, I'm going always to "Article" form. I neither have the choice between Article and Page.
1. If I "Add" the reference in new form, I'm going to the front display, not going back to original form.
1. If I edit a content with an existing reference then press "Edit" I'm going back to the listing content page.
1. If I want to add a value into a field which already have a reference, I'm going back to the listing content page.

I will give a try with a MySQL connected instance to eliminate an issue with SQLite.

Will back with more tests soon.

**EDIT:**
- Comment still valid with MySQL.
- New issue added

**Comment by jygastaud** [comment]
hum maybe I misread the fact that PR is only for buttons displays, not for more.
If it's that, I will give a try more specifically on that.

@amaria can you confirm me the scope I need to test ?

**Comment by amaria** [comment]
Yes, only on button displays. They don't actually do anything yet.

**Comment by jygastaud** [comment]
Hi @amaria,

Here is my feedback for buttons display :
- Node Reference : Everything looks good for the 3 buttons
- User Reference : _Add_ is not display for both limited and unlimited 
- Taxonomy : Look fine

File: src/EntityconnectWidgetProcessor.php [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+      if (\Drupal::entityTypeManager()
+        ->getAccessControlHandler($this->entityType)
+        ->createAccess($bundle)
+      ) {
+        $acceptableTypes[] = $bundle;
+      }
+    }
+
+    // Now we need to make sure the user should see this button.
+    if (\Drupal::currentUser()->hasPermission('entityconnect add button') && $addbuttonallowed && $acceptableTypes) {
```
Comment (jygastaud) [inline src/EntityconnectWidgetProcessor.php:228]: `$acceptableTypes` cause the issue in user reference as there is no option for type in User reference.

**Comment by jygastaud** [comment]
To not lock anymore that PR, I merge it and create a separated issue.

**Comment by jygastaud** [comment]
Thanks for your work @amaria

**Comment by amaria** [comment]
Thanks. Sorry I haven't been too responsive on this. I'm working on a paid project so haven't had a lot of time for the contributed work lately.

**Pull request merged**